src/event_loops_bridge.rs

```diff
@@ -117,3 +119,38 @@ fn pop_scheduled_outside_dlg() -> Option<Box<dyn FnOnce()>> {
         queue.pop_front()
     })
 }
+
+pub async fn wake_from<T>(setup_dlg: impl FnOnce(oneshot::Sender<T>)) -> Result<T, oneshot::error::RecvError> {
+    let (tx, rx) = oneshot::channel();
+    setup_dlg(tx);
+    rx.await
+}
+
+pub async fn outside<T: 'static>(fut: impl Future<Output = T> + 'static) -> Result<T, oneshot::error::RecvError> {
+    wake_from(|tx| {
+        glib::MainContext::ref_thread_default().spawn_local(async move {
+            let result = fut.await;
+            tx.send(result).map_err(|_| "Unable to send future result").unwrap();
+        });
+    })
+    .await
+}
+
+pub async fn run_dialog(
+    dialog: &(impl gtk::DialogExt + gtk::GtkWindowExt + gtk::WidgetExt),
+    close_after: bool,
+) -> gtk::ResponseType {
+    dialog.set_modal(true);
+    dialog.show();
+    wake_from(|tx| {
+        let tx = std::cell::Cell::new(Some(tx));
```

Comment:
I'm not happy with always having to use a cell here. I think about making this the "raw" function (renaming it to `wake_from_sender`) and have the commonly used `wake_from` accept a closure that wraps over this cell.